### PR TITLE
Datasources: Simplify the AzureCredentials structure in datasource config

### PR DESCRIPTION
**Pull request by kostrse**
**What this PR does / why we need it**:

In the initial experimental implementation of Azure authentication in `httpclientprovider` of plugin proxy, there're two fields which should be set: `azureAuth: boolean` flag and `azureCredentials : AzureCredentials` structure.

`azureAuth` was controlled by a UI toggle and `azureCredentials` contained the actual credentials configuration.
`azureAuth` flag had to be set to true to `azureCredentials` take effect.

Having two fields implies extra logic in Grafana `httpclientprovider` code more room for error. Also it requires datasources to handle both of the fields.

This PR removes necessity for `azureAuth`. if `azureCredentials` then Azure authentication is enabled, otherwise it not. Datasources have to set/unset only one field.

**Which issue(s) this PR fixes**:

Related #35857 

**Special notes for your reviewer**:

The changes affect `httpclientprovider` (plugin proxy) and Prometheus datasource.

`DatasourceHttpSettings` don't know about specifics of `azureAuth` toggle anymore, everything is abstracted out via callbacks.

All of this Azure authentication functionality is currently under experimental flags and doesn't affect existing Grafana features.

### CODE REVIEW

**Comment by kostrse** [comment]
@joshhunt could you please have a look at this refactoring PR?
All the tests fixed,   only `test-a11y-frontend-pr` failed which I see is failing for everyone.

**Review by joshhunt — COMMENTED** [review]
Looks good to me, at least the frontend part of it. I don't have an environment available to test this.

Will defer to others for the final approval for the rest of it, esp. with the proxy auth.

**Comment by kostrse** [comment]
Currently this logic used only by the Prometheus datasource and it's under experimental feature toggles.

### How to test it

1. Enable the feature toggles:
```ini
[feature_toggles]
enable = prometheus_azure_auth,httpclientprovider_azure_auth
```
2. Create a new Prometheus datasource
3. Use a mock endpoint (e.g. [RequestBin](https://requestbin.com) or [Mockoon](https://mockoon.com) etc.) for the created Prometheus datasource.
4. Click **Save & Test**, verify that the request was sent without an `Authorization` header.
5. Enable switch **Azure Authorization** and configure **App Registration** settings.
6. Set value of **AAD resource ID** to `https://management.azure.com`
image
7. Click **Save & Test** again, verify that the request was sent with an `Authorization` header and the token was issued for the `https://management.azure.com` audience (https://jwt.ms/).
image
image
8. Disable the switch **Azure Authorization**
9. Click **Save & Test** again, verify that the request doesn't contain the `Authorization` header anymore.

File: pkg/models/datasource_cache.go
Comment (kostrse) [reply pkg/models/datasource_cache.go]: Thanks for the feedback, @marefr

I saw that the `datasource_cache` used only by plugin proxy and Prometheus plugin is transitioning to implementation in Plugin SDK (e.g. [here](https://github.com/grafana/grafana-plugin-sdk-go/blob/4c305c4d10215404291b5c2963a83f7acd558264/backend/http_settings.go#L128)) and things like "test connection" data dashboards already use that implementation.

It needs to be implemented in Plugin SDK and I will send a separated PR for that. And we would discuss that.

I would want to merge this PR before because **this PR doesn't introduce any new design concepts or assumptions** but it cleans up the schema of Azure credentials in JsonData and removes unnecessary knowledge of Azure credentials schema from the `DatasourceHttpSettings` component.

This would be very helpful before further discussion on support of Azure credentials [here](https://github.com/grafana/grafana-plugin-sdk-go/blob/4c305c4d10215404291b5c2963a83f7acd558264/backend/http_settings.go#L128) etc.

Right now this is under experimental feature flag `httpclientprovider_azure_auth` in Plugin SDK.


#### `azureCredentials` vs `_azureCredentials`

The original JSON object that comes from UI is `azureCredentials`. But that object is a raw set of key/values which are not validated. `_azureCredentials` holds  `AzureCredentials`, a domain object, abstract from how it's being serialized and stored by a plugin.

`datasource_cache` (as a builder for `httpclient.Options`), validates raw user input and deserializes `AzureCredentials` to `CustomOptions["_azureCredentials"]`.

So, `azureCredentials` is RAW key/value pairs received from frontend, `_azureCredentials` is a parsed `AzureCredentials` object.

Azure middleware (and Azure Token Provider) works with `AzureCredentials`.

For example, Azure Monitor has own, already established, JSON schema for Azure credentials and it has own implementation of parsing it from JsonData settings. Having `AzureCredentials` in `httpclient.Options`  allows to abstract out of variations in JsonData schemas of different plugins, and reuse the same generic Azure middleware implemented in `httpclient`/`httpclientprovider`.

For example, Azure Monitor could just provide credentials [here](https://github.com/grafana/grafana/blob/96efbbaed1728e89602ca9578a24d24bd2ceb53c/pkg/tsdb/azuremonitor/httpclient.go#L42-L41) to `httpclient.Options` instead of implementing and passing own middleware:

```go
func newHTTPClient(route azRoute, model datasourceInfo, cfg *setting.Cfg, clientProvider httpclient.Provider) (*http.Client, error) {
	m, err := getMiddlewares(route, model, cfg)
	if err != nil {
		return nil, err
	}

	return clientProvider.New(httpclient.Options{
                AzureCredentials: model.Credentials, // CustomOptions["_azureCredentials"]
		Headers:          route.Headers,
		Middlewares:      m,
	})
}
```

Comment (marefr) [reply pkg/models/datasource_cache.go]: Hey @kostrse. Sorry for the somewhat late response. 

We've had some internal discussions and there's a need that have arise  for enable azure authentication for azure data sources not builtin to Grafana, e.g. external ones. Historically when we discussed this everything was experimental, but now feels like a more appropriate time for considering adding things to the grafana-plugin-sdk-go repository.

> I would want to merge this PR before because **this PR doesn't introduce any new design concepts or assumptions** but it cleans up the schema of Azure credentials in JsonData and removes unnecessary knowledge of Azure credentials schema from the `DatasourceHttpSettings` component.
> 
> This would be very helpful before further discussion on support of Azure credentials [here](https://github.com/grafana/grafana-plugin-sdk-go/blob/4c305c4d10215404291b5c2963a83f7acd558264/backend/http_settings.go#L128) etc.

Makes sense. And it seems like you're already planning to go in the direction I'm suggesting below after this has been merged :+1: I'll go ahead and review this again so we can get it merged. I think I have your email since earlier meeting so I'll just send an email in case we need to sync up on things.

> It needs to be implemented in Plugin SDK and I will send a separated PR for that. And we would discuss that.

Great. We would like the SDK to include the settings/properties for azure authentication similar to the AWS SigV4 settings as can be seen [here](https://github.com/grafana/grafana-plugin-sdk-go/blob/329c72e4e242f84c3b185013f4863fcce7996086/backend/http_settings.go#L39-L46), but keep the SDK as agnostic as possible. With that said, it would be good to keep the shared logic for azure authentication in a separate repository so that Grafana and external plugins could make use of it. Seems like there's an existing repository created, [grafana-azure-sdk-go](https://github.com/grafana/grafana-azure-sdk-go), which seems like a good fit. We have this setup for AWS as inspiration/guidance, [grafana-aws-sdk](https://github.com/grafana/grafana-aws-sdk).

Comment (marefr) [reply pkg/models/datasource_cache.go]: Seems like CI refuses to start so you'll probably have to merge with latest changes from main to get it running.

**Review by marefr — COMMENTED** [review]
Think the changes looks good. Would like to take it for a quick spin when CI build is green.

File: packages/grafana-ui/src/components/DataSourceSettings/types.ts
Comment (mckn) [inline packages/grafana-ui/src/components/DataSourceSettings/types.ts:6]: I think this initiative is great and I agree that we need to improve the code according to your description in the PR.

However, this is a breaking change and will make plugins using these APIs to stop functioning until being upgraded to use this version of `@grafana/ui`. 

Is this used only by bundled plugins or do we have any external plugins using this? @grafana/cloud-datasources do you know?

**Review by mckn — COMMENTED** [review]
I think this is great! But we should probably do some refactoring of this code so it gets a bit more generic and introduce it as a non-breaking change.

File: packages/grafana-ui/src/components/DataSourceSettings/types.ts
Comment (sunker) [reply packages/grafana-ui/src/components/DataSourceSettings/types.ts:6]: Not that I know of, but theoretically it can be used by any data source that enable this feature. I know for sure that the only other Azure data source that we maintain - Azure data explorer - authenticates directly against Azure so it's not using the http client provider. @kostrse do you know if any other data sources that might be using `Azure auth` through DataSourceHttpSettings?

Comment (kostrse) [reply packages/grafana-ui/src/components/DataSourceSettings/types.ts:6]: This has been used only for Prometheus datasource and behind a feature flag (`prometheus_azure_auth`).
So, while it's a change to public interface, there should not be any impact.

**Comment by kostrse** [comment]
> I think this is great! But we should probably do some refactoring of this code so it gets a bit more generic and introduce it as a non-breaking change.

@mckn @sunker @marefr Please, what else needed to have it finally merged?
This functionality only used in Prometheus plugin _behind experimental feature flag_. This is not a breaking change

**Comment by mckn** [comment]
> > I think this is great! But we should probably do some refactoring of this code so it gets a bit more generic and introduce it as a non-breaking change.
> 
> @mckn @sunker @marefr Please, what else needed to have it finally merged? This functionality only used in Prometheus plugin _behind experimental feature flag_. This is not a breaking change

Okay, super! Then it should be fine to proceed with the changes.

**Review by mckn — APPROVED** [review]
LGTM

**Comment by sunker** [comment]
Looks good to me too! @marefr can we merge this know?

**Review by marefr — APPROVED** [review]
LGTM. I merged with main to resolve conflicts and pushed the changes.

**Comment by marefr** [comment]
Thanks @kostrse. Let us know how you would like to proceed with the next steps discussed in comments above.

**Comment by kostrse** [comment]
> Thanks @kostrse. Let us know how you would like to proceed with the next steps discussed in comments above.
@marefr  Yes. I will continue sending follow-up PRs.

Now I have PR #44407 which fixes Azure authentication in Prometheus datasource after it was regressed in 8.3 due to refactoring of Prometheus from `plugin.json` routes to httpclient in Go backend. Now it will work for Alerts too.

**Pull request merged**
